nbodykit/algorithms/pair_counters/corrfunc/base.py

```diff
@@ -0,0 +1,193 @@
+import numpy
+import logging
+from nbodykit import CurrentMPIComm
+from nbodykit.binned_statistic import BinnedStatistic
+
+class MissingCorrfuncError(Exception):
+    def __init__(self):
+        msg = "use either ``conda install -c bccp corrfunc`` "
+        msg += "or ``pip install git+git://github.com/nickhand/Corrfunc``"
+        self.args = (msg,)
+
+class CorrfuncResult(object):
+    """
+    A class used internally to hold the array-like result of a
+    pair counting algorithm from :mod:`Corrfunc`.
+
+    This class is useful for reducing pair count results in parallel
+    while accounting for columns that are pair-weighted.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        the numpy structured array result from :mod:`Corrfunc`
+    """
+    valid = ['weightavg', 'npairs', 'savg', 'ravg', 'thetaavg', 'rpavg']
+
+    def __init__(self, data):
+
+        # copy over the valid colums from the input result
+        dtype = [(col, data.dtype[col]) for col in data.dtype.names if col in self.valid]
+        self.data = numpy.zeros(data.shape, dtype=dtype)
+        self.columns = self.data.dtype.names
+        for col in self.columns:
+            self.data[col] = data[col]
+
+    def __getitem__(self, col):
+        return self.data[col]
+
+    @property
+    def shape(self):
+        return self.data.shape
+
+    @property
+    def dtype(self):
+        return self.data.dtype
+
+    def reshape(self, *args, **kwargs):
+        self.data = self.data.reshape(*args, **kwargs)
+        return self
+
+    def __radd__(self, other):
+        return self + other
+
+    def __add__(self, other):
+
+        # all columns are pair-weighted except for "npairs"
+        data = numpy.empty_like(self.data)
+        for col in self.columns:
+            if col == 'npairs': continue
+            data[col] = self[col]*self['npairs'] + other[col]*other['npairs']
+
+        # just sum up "npairs"
+        data['npairs'] = self['npairs'] + other['npairs']
+
+        # normalize by total "npairs"
+        idx = data['npairs'] > 0.
+        for col in self.columns:
+            if col == 'npairs': continue
+            data[col][idx] /= data['npairs'][idx]
+
+        return CorrfuncResult(data)
+
+class MPICorrfuncCallable(object):
+    """
+    A base class to represent MPI-enabled Corrfunc callable.
```

Comment:
Perhaps a link or more text what these callables are used for.

Reply:
sure no problem

Reply:
One of the things that these classes do are try to handle exceptions from Corrfunc sensibly, since it's a bit broken at the moment (as you've noted). We capture stdout/stderr and the C-level logging, and if the Corrfunc calls, print it all to screen. 

There seems to be helpful C-level logging for errors in Corrfunc, but the Python exception raised is not helpful.